[wso2iots] WSO2 IoT Server Binding initial contribution

RameshaKaru opened the pull request:
"[wso2iots] INITIAL CONTRIBUTION NEW BINDING"

This is a novel addition of a Binding named 'wso2iots', which is used to connect openHAB2 with WSO2 IoT Server 3.1.0 (https://wso2.com/iot) and get temperature, humidity, light level, motion data from devices enrolled in the building monitor plugin of WSO2 IoTS. Further detatils are included in wso2iots addon README.md
Signed-off-by: Ramesha Karunasena ramesha.karu@gmail.com (github: github_handle)

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/OSGI-INF/.gitignore:
```diff
@@ -0,0 +1 @@
+*.xml
```
I'm not sure whether bindings should contain their own `.gitignore` files

martinvw replied:
For this directory it should 😔 At least for now

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+# wso2iots Binding
```
wso2iots -> WSO2 IoT Server

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+		Text item=wso2_Motion label="Motion [%.1f ]" icon="motion"
+		Text item=wso2_Humidity label="Humidity [%.1f %%]" icon="humidity"
+		Text item=wso2_Light label="Light [%.1f ]" icon="light"
+		
```
Trailing newline

ppieczul replied:
remove empty lines

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+The information that is retrieved is available as these channels:
+
+
+| Channel ID | Item Type    | Description              |
```
Please reformat the table. See www.tablesgenerator.com/markdown_tables

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/.idea/misc.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please remove intellij-specific stuff. BTW what is the reason for using intellij vs Eclipse? If this is just a matter of preference, I would suggest to keep it outside of this repo, as openHAB supports eclipse. Having two will most likely cause one or the other to become obsolete - for example other community users making changes will not update the intellij.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+
+	<name>wso2iots Binding</name>
+	<description>This is the binding for wso2iots.</description>
+	<author>wso2_Ramesha</author>
```
Please use full first and last name

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/i18n/wso2iots_xx_XX.properties:
```diff
@@ -0,0 +1,13 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
do or remove localization files

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="wso2iots"
+						  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
fix justification of the whole file, so it displays correctly here

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+						  xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+						  xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
remove this comment

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+
+	<!-- Sample Thing Type -->
+	<thing-type id="buildingMonitor">
+		<label>Building Monitor Data</label>
```
The thing seems to be Building Monitor, not Data

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+
+	</thing-type>
+
+	<!-- Sample Channel Type -->
```
remove this comment

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+
+	<!-- Sample Channel Type -->
+	
+	<channel-type id="temperature" advanced="true">
```
why these channels are advanced?

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+		
+	}
+
+
```
Trailing newlines

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+import org.apache.commons.lang.StringUtils;
+//import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+//import org.eclipse.smarthome.core.library.types.StringType;
```
Please remove commented out code

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+		<item-type>Number</item-type>
+		<label>Temperature (°C)</label>
+		<description>Temperature in Celsius degrees</description>
+		<category>Temperature (°C)</category>
```
If there is no another category in F, probably this should be just Temperature

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+            if (channelUID.getId().equals(wso2iotsBindingConstants.MOTION)) {
+                updateChannel(channelUID.getId(), iotsResponseMotion);
+            }
+            if (channelUID.getId().equals(wso2iotsBindingConstants.HUMIDITY)) {
```
Wouldn't `switch` fit here better?

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+	<channel-type id="temperature" advanced="true">
+		<item-type>Number</item-type>
+		<label>Temperature (°C)</label>
+		<description>Temperature in Celsius degrees</description>
```
Is the unit system defined by the device? Or this is always celsius? Do we have a story for overseas?

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+            in.close();
+
+            // print result
+            // System.out.println(response.toString());
```
Commented out code

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+		<label>motion</label>
+		<description></description>
+		<category>motion</category>
+		<state readOnly="true" pattern="%.1f "></state>
```
you can use /> closing

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonMotion/Values.java:
```diff
@@ -0,0 +1,28 @@
+ * The {@link Values} class returns the motion value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name here

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+
+	<channel-type id="motion" advanced="true">
+		<item-type>Number</item-type>
+		<label>motion</label>
```
Consistency in naming convention: capital or small letters, please unify across this xml for labels, descriptions and categories.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+	<channel-type id="humidity" advanced="true">
+		<item-type>Number</item-type>
+		<label>Humidity level</label>
+		<description></description>
```
Please add description

kubawolanin commented on addons/binding/pom.xml:
```diff
@@ -94,5 +93,6 @@
     <module>org.openhab.binding.yamahareceiver</module>
     <module>org.openhab.binding.zoneminder</module>
     <module>org.openhab.binding.zway</module>
+    <module>org.openhab.binding.wso2iots</module>
```
Please follow the alphabetical order.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,81 @@
+	<channel-type id="light" advanced="true">
+		<item-type>Number</item-type>
+		<label>light</label>
+		<description></description>
```
Please add description

kubawolanin commented on addons/binding/pom.xml:
```diff
@@ -25,7 +25,6 @@
     <module>org.openhab.binding.bigassfan</module>
     <module>org.openhab.binding.boschindego</module>
     <module>org.openhab.binding.chromecast</module>
-    <module>org.openhab.binding.cm11a</module>
```
Why is this removed?

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/wso2iotsBindingConstants.java:
```diff
@@ -0,0 +1,40 @@
+ * The {@link wso2iotsBindingConstants} class defines common constants, which are
+ * used across the whole binding.
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/wso2iotsHandlerFactory.java:
```diff
@@ -0,0 +1,54 @@
+ * The {@link wso2iotsHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/wso2iotsConfiguration.java:
```diff
@@ -0,0 +1,25 @@
+
+/**
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonTemp/wso2iotsResponseTemp.java:
```diff
@@ -0,0 +1,27 @@
+ * The {@link wso2iotsResponseTemp} class returns the temperature value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonTemp/Values.java:
```diff
@@ -0,0 +1,29 @@
+ * The {@link Values} class returns the temperature value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonMotion/wso2iotsResponseMotion.java:
```diff
@@ -0,0 +1,27 @@
+ * The {@link wso2iotsResponseMotion} class returns the motion value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonLight/wso2iotsResponseLight.java:
```diff
@@ -0,0 +1,27 @@
+ * The {@link wso2iotsResponseLight} class returns light value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonLight/Values.java:
```diff
@@ -0,0 +1,28 @@
+ * The {@link Values} class returns light value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonHumidity/wso2iotsResponseHumidity.java:
```diff
@@ -0,0 +1,27 @@
+ * The {@link wso2iotsResponseHumidity} class returns the humidity value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonHumidity/Values.java:
```diff
@@ -0,0 +1,29 @@
+ * The {@link Values} class returns the humidity value
+ * from json wso2iotsresponse string
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use your real name

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+            errormsg = e.getMessage();
+            logger.warn(errormsg);
+            resultok = false;
+            // return null;
```
Commented out code

ppieczul replied:
cleanup

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                    iotsResponse = result.getvalues().getData();
+                }
+                resultok = true;
+                // return iotsResponse;
```
Commented out code

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+
+WSO2 is an opensource middleware company and IoTS is one of the five products offered by the company.(https://wso2.com/)
+
+To use the binding, first the user have to locally setup the WSO2 IoTS (https://wso2.com/iot) and deploy the building monitor plugin (https://github.com/wso2/samples-iots/tree/master/floor-analytics-demo). The building monitor agent uses ESP8266 12E, DHT 11, PIR SR501, LM393. (https://github.com/wso2/samples-iots/tree/master/floor-analytics-demo/BuildingMonitorDevice). 
```
have -> has

ppieczul replied:
It would look more smooth, if you hide urls in links (e.g. [the building monitor plugin](https://github.com/wso2/samples-iots/tree/master/floor-analytics-demo))

ppieczul replied:
uses ... LM393 devices

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                resultok = false;
+                errormsg = "empty response from wso2 iots";
+                logger.warn(errormsg);
+                // return null;
```
Commented out code

ppieczul replied:
cleanup

kubawolanin commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+            int responseCode = con.getResponseCode();
+
+            // logger.debug("\nSending 'GET' request to URL : " + url);
+            // logger.debug("Response Code : " + responseCode);
```
Commented out code

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## Supported Things
+
+There is exactly one supported thing type, which represents one device agent. It has the `buildingMonitor` id. Of course, you can add multiple Things, e.g. Devices enrolled in different floors and different buildings.
```
'Of course...' - seems redundant in the readme file.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## Thing Configuration
+
+The thing has a few configuration parameters:
```
a few -> the following (?)

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+| motion | Number | Motion level in the range 0-1 |
+| temperature | Number | Temperature in Celsius degrees |
+| light | Number | Light level in the range 0-1024|
+| humidity | Number | Humidity level |
```
what units? (consistency)

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/README.md:
```diff
@@ -0,0 +1,87 @@
+| light | Number | Light level in the range 0-1024|
+| humidity | Number | Humidity level |
+
+Note that light level 0 mean high brightness and 1024 means very dark
```
Move this to the table and add similar explanation to the other rows.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/org.openhab.binding.wso2iots.iml:
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
remove intellij specific file

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+ * The {@link wso2iotsHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author wso2_Ramesha - Initial contribution
```
Please use full first and last name (everywhere)

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+ *
+ * @author wso2_Ramesha - Initial contribution
+ */
+// @NonNullByDefault
```
It would be nice to enable this actually (or remove)

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+     * @Override
+     * public void handleCommand(ChannelUID channelUID, Command command) {
+     * if (channelUID.getId().equals(CHANNEL_1)) {
+     * // TODO: handle command
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+     */
+    @Override
+    public void initialize() {
+        // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+
+        logger.debug("Initializing wso2iots handler.");
+
+        wso2iotsConfiguration config = getConfigAs(wso2iotsConfiguration.class);
```
Change name to comply with class naming convention (CamelCase)

ppieczul replied:
Still TBD

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+            errorMsg = "Parameter 'deviceId' is mandatory and must be configured";
+            validConfig = false;
+        }
+        if (config.refresh != null && config.refresh < 2) {
```
no logging when refresh is null

RameshaKaru replied:
If refresh is null it takes DEFAULT_REFRESH_PERIOD as refresh period in the startAutomaticRefresh method

ppieczul replied:
but this should never occur and would mean some problem, the refresh period will always be passed on from the configuration, because it has a default value in the xml

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        }
+
+        if (validConfig) {
+            updateStatus(ThingStatus.ONLINE);
```
It seems to early to set the status to online, when no communication attempt was made yet.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                updateChannel(channelUID.getId(), iotsResponseHumidity);
+            }
+        } else {
+            logger.debug("The Building Monitor binding is read-only and can not handle command {}", command);
```
No clear what the remark about binding being read-only means.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                        updateChannel(wso2iotsBindingConstants.MOTION, iotsResponseMotion);
+                        updateChannel(wso2iotsBindingConstants.HUMIDITY, iotsResponseHumidity);
+
+                        // for (Channel channel : getThing().getChannels()) {
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                }
+            };
+
+            wso2iotsConfiguration config = getConfigAs(wso2iotsConfiguration.class);
```
You can store the parameters locally, so you don't parse the config to the class at each refresh. When new configuration appears, the binding will be disposed and initialized again.

RameshaKaru replied:
I thought if binding is not initialized every time the configurations change and if a long time period is given as the refresh interval (ex: 1 hour), then the user wont be able to know until 1 hour whether the data is received properly after configuration is changed. Therefore is it better to initialize the binding each time the configurations change?

ppieczul replied:
The framework will call initialize after each configuration change for you

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+     */
+    private void updateChannel(String channelId, BigDecimal iotsResponse) {
+        if (isLinked(channelId)) {
+            /*
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+    /**
+     * Update the channel from the last building monitor data retrieved
+     *
+     * @param channelId the id identifying the channel to be updated
```
Please run javadoc and cleanup all errors/warnings

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                logger.warn("Update channel {}: Unsupported value type {}", channelId,
+                        iotsResponse.getClass().getSimpleName());
+            }
+            logger.debug("Update channel {} with state {} ({})", channelId, (state == null) ? "null" : state.toString(),
```
it should actually work if you pass even null arguments to the logger, it will print null

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        String fromTime = String.valueOf(time - 200);
+        String toTime = String.valueOf(time);
+
+        String url = "https://localhost:8243/senseme/device/1.0.0/stats/";
```
Could you add an ability to query any server not just localhost - this seems like a limitation simple to overcome.

RameshaKaru replied:
my idea is to have the WSO2 IoT Server locally hosted and connect with openHAB

ppieczul replied:
Would that be suitable for all users? What is the reasoning for limiting this just to the localhost?

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        boolean resultok = false;
+        String errormsg = null;
+        String response1 = null;
+        wso2iotsConfiguration config = getConfigAs(wso2iotsConfiguration.class);
```
same comment as previously, so actually this happens for each query at each refresh

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        wso2iotsConfiguration config = getConfigAs(wso2iotsConfiguration.class);
+        String Access_Token = config.apikey;
+        String id = config.deviceId;
+        // String Access_Token = "bfe52f8a-5f34-382b-802c-93e52468564f";
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        // String Access_Token = "bfe52f8a-5f34-382b-802c-93e52468564f";
+        // String id = "2940205";
+        long time = System.currentTimeMillis() / 1000;
+        String fromTime = String.valueOf(time - 200);
```
why 200? should this be a configuration parameter? otherwise, please use a constant

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        url = url + id + "?from=" + fromTime + "&to=" + toTime + "&sensorType=" + sensorType;
+
+        URL obj;
+        // if(Boolean.parseBoolean(System.getProperty(""))) {
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+
+            int responseCode = con.getResponseCode();
+
+            // logger.debug("\nSending 'GET' request to URL : " + url);
```
cleanup

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+                logger.debug(segment);
+                // Map the JSON response to an object
+
+                if (sensorType == wso2iotsBindingConstants.TEMPERATURE) {
```
Is it possible to query all sensors altogether?

RameshaKaru replied:
building monitor plugin has the API to get sensor data separately

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        } catch (Exception e) {
+            errormsg = e.getMessage();
+            logger.warn(errormsg);
+            resultok = false;
```
you should be able to infer resultok from the errormessage == null

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,390 @@
+        // if(Boolean.parseBoolean(System.getProperty(""))) {
+        obj = this.validateCertificate(url);
+
+        HttpURLConnection con = (HttpURLConnection) obj.openConnection();
```
what if validation failed or server is not reachable?

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonTemp/wso2iotsResponseTemp.java:
```diff
@@ -0,0 +1,27 @@
+/**
```
if responses for each sensor hangs under different json branch, you should be able to actually use a single class to deserialize this data, with all branches in it. gson will fill in the values which it will find in the incoming json.

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/wso2iotsHandlerFactory.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @author wso2_Ramesha - Initial contribution
+ */
+@Component(service = ThingHandlerFactory.class, immediate = true, configurationPid = "binding.wso2iots")
```
you can skip configurationPid, as you don't have binding configuration

ppieczul commented on addons/binding/pom.xml:
```diff
@@ -25,7 +25,6 @@
     <module>org.openhab.binding.bigassfan</module>
```
You also need to add your binding to the feature.xml file.

RameshaKaru commented:
Thank you very much for the feedback. I will do the suggested the changes

ppieczul commented:
One more thing, probably might be the key here. Since what this binding does is getting a JSON from a web server and then parsing it, it should be possible to use the [JsonPath](http://docs.openhab.org/addons/transformations/jsonpath/readme.html) transformation combined with http binding. Could you give it a try and see if it is possible and is it easy?
This would mean this binding is not needed. I am not really a big fan of transformations instead of dedicated bindings, because they need a higher level of technical knowledge to configure. I have a similar problem with a wifi cesspool sensor, that exposes its data over http in xml format. It was easier for me to write a simple binding than to get myself familiar with xslt transformations, which I have never worked with. Anyway, what may come out of it, is a need to make transformations more user-friendly. My thoughts go now in the direction of a kind of a marketplace or addons pool that would cover the transformations too.

RameshaKaru commented:
This binding is written only to handle one thing (building monitor) at the moment. However I am planning to develop this binding by adding few other plugins in WSO2 IoTS as things and to separately handle those thing types too. That is why I decided to create a binding to connect the two servers.

RameshaKaru commented:
Thanks for the feedback and I have done the suggested changes and committed it

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -50,16 +46,18 @@
 
     private static final int DEFAULT_REFRESH_PERIOD = 60;
+    
+    private static final int PERIOD = 200;
```
A bit more verbose name would help to understand what this is

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -241,41 +203,32 @@ private void updateChannel(String channelId, BigDecimal iotsResponse) {
-        // optional default is GET
-        con.setRequestMethod("GET");
+        
+        try {
```
For validateCertificate, moving try before it does not change anything, because you catch the exceptions inside that method. Question remains valid - can validateCertificate fail and if yes, what are the consequences to the code following it - will it fail? How will it fail?

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonTemp/Values.java:
```diff
@@ -15,7 +15,7 @@
+ * @author Ramesha Karunasena - Initial contribution
  */
 
 public class Values {
```
The previous comment remains valid - I would suggest combining all json classes into single class, because it leads to a much simpler code which is easier to understand and maintain. Unless there are some reasons not to :)

ppieczul commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -43,39 +42,37 @@
-		<category>Temperature (°C)</category>
-		<state readOnly="true" pattern="%.1f °C"></state>
+		<category>Temperature</category>
+		<state readOnly="true" pattern="%.1f °C" />
```
Still valid comment - how does the server handle celsius vs. fahrenheit?

RameshaKaru commented:
I combined the 4 json classes into one class, removed the default refresh period. Updated the channel to change to offline if validateCertificate() fails. For the temperature, the API sends data only in Celsius.
Thanks a lot for the suggested changes.

ppieczul commented:
Could you help me understand on a high level how the wso2iot works wrt your code. From wso2iot web page I understand there is a wso2iot server, which user can install. This server offers supervision of various devices through plugins. 

What is in this context "senseme" and its sensors you query - is it one of many available device plugins? If yes, who supports it and where can I find it on the web and try it out?

RameshaKaru commented:
WSO2 IoT Server allows the user to deploy different plugins to the server and building visualizer/monitor is one such plugin. senseme is the device agent of the building visualizer plugin. 

Building visualizer plugin allows the user to add buildings in a map using a UI. User can give the floor map of the building and add the senseme device agents on the floor map. When the senseme device sends data to the WSO2 IoTS, they are displayed as graphs and also WSO2 analytics is used to analyze them and create notifications for very low temperatures etc and to get info like average temperature for the floor.

The wso2iots binding allows the user to integrate openHAB with WSO2 IoT Server and visualize the data sent from senseme devices in WSO2 IoT server. Currently the binding facilitates to connect with only the senseme device type. However, I am planning to add the device types of few other plugins in the WSO2 IoT Server and handle them as separate Thing types.

Different plugins supported in WSO2 IoT Server can be found [here](https://github.com/wso2/samples-iots)
Building Visualizer, the updated version can be found [here](https://github.com/RameshaKaru/floor-analytics-demo)
The documentation on how to configure the building visualizer plugin and how to make the hardware part is still on progress. I will send you the documentation once it is completed.

Thanks

ppieczul commented:
Thank you very much for the explanations. Based on this wso2iot architecture I would like to discuss changes to the structure of your binding to reflect that architecture and allow for your planned evolution to support more device. 
  * Separation of wso2iot generic stuff from the device/agent specific stuff, so support for more devices can be easily added.
    * If your design is to have different thing types for different devices (which I think makes sense), than the currently implemented thing would be named and maybe also packaged following the device name (senseme), not wso2iot, which is the generic server.
    * How do you abstract the server-specific stuff from the device-specific (e.g. token, server host, API paths)? You could consider making an OH bridge per server which has things for the devices. Even if not implementing the bridge, the server-specific stuff should be separated from the device-specific in the code, so it is not repeated for all future devices and maintained in a single code place.
    * Accessing device API - the request URL `https://localhost:8243/senseme/device/1.0.0/stats/` - similarly separation of management of parts of that URL that are server-specific from that device-specific (not sure which are which). How do we manage the 1.0.0 part - it seems like a version - so it is not hard-coded? Is there an API a server part could use to query the devices and their versions, so they are always properly accessed even if they change on the server? Also I think the configurable server host name and port is a must, it should not be limited to the localhost without a strong reason.
    * Processing API responses - who defines the JSON response structure? The parsing logic should follow the separation from server and device, so in case the server API definition changes, it can be updated in a single place and not in each device thing handler.
Looking forward to your opinion about that.
Thanks
Pawel

RameshaKaru commented:
Thanks for your valuable feedback 
1. I have represented the senseme device type by the Building Monitor Thing type. Therefore, I will change the handler name from wso2iotshandler to BuildingMonitorhandler
2. I considered adding a bridge. But since the access token may differ for the different device types, I did not add it. Therefore shall I separate the server specific parts into a separate class without implementing a bridge? What will be the better approach?
3. I will add the port and server host name to thing configuration parameters. 
There isn't an API to query the versions of the available device types. Therefore, I think the version has to be hard-coded.
4. JSON response structure is defined for each device type.

The steps to deploy and use the building monitor/visualizer plugin can be found [here](https://medium.com/@ramesha.karu/wso2-iots-plugin-building-visualizer-ba603756953d)

Thanks

RameshaKaru commented:
Further for the 3rd problem about API version- The WSO2 API Management supports the functionality of default version, so that the API does not have to include the version and when the API is called it automatically choose the default version. But currently this feature is not implemented in WSO2 IoT Server. I have raised a git issue in wso2 iots repository: https://github.com/wso2/product-iots/issues/1457

For the 4th problem, since the JSON response structure is defined for each device type and device type is associated with a plugin, the dependency between the JSON response structure and the plugin cannot be decoupled. Therefore, if the JSON structure changes (which means the API has changed) the binding has to be modified manually. 

Thank you

ppieczul commented:
@RameshaKaru thank you for all the explanations. I suppose for the bridge, it makes sense to add it if there are configuration parameters that are common for all devices. Additional pro argument would be a future ability to implement autodiscovery of the devices, if this is ever possible with this server. However, if authentication token is really proprietary to the senseme device and not wso2iot server, it may be fine without a bridge. I would leave this up to your decision, given this as an input.

RameshaKaru commented:
Thank you. I will keep the binding without a bridge for the moment, since there is only one thing type currently and I will add a bridge when I add more thing types in future.

martinvw commented:
> Thank you. I will keep the binding without a bridge for the moment, since there is only one thing type currently and I will add a bridge when I add more thing types in future.

That would make extending the binding later a breaking change, because existing users will need reconfigure their openHAB items and things. So I would not be happy with that decision.

RameshaKaru commented:
@martinvw , I did not consider that point earlier. Thanks a lot for showing it. I am tight on schedule at the moment. Therefore I will add the bridge and will update the thread on next weekend.
Thanks

Hilbrand commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,334 @@
+            State state = null;
+            if (iotsResponse == null) {
+                state = UnDefType.UNDEF;
+            } else if (iotsResponse instanceof BigDecimal) {
```
iotsResponse **is** a BigDecimal so why have an else? It will only return false if the value is null. Therefor the else can actually be the test above:
```
if (iotsResponse instanceof BigDecimal) {
  ...
} else {
  state = UnDefType.UNDEF;
}
```

Hilbrand commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,334 @@
+        String fromTime = String.valueOf(time - API_RESPONSE_INTERVAL);
+        String toTime = String.valueOf(time);
+
+        String url = "https://localhost:8243/senseme/device/1.0.0/stats/";
```
Does this always run on localhost? else it should be a configuration option.

Hilbrand commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/wso2iotsHandler.java:
```diff
@@ -0,0 +1,334 @@
+                while ((inputLine = in.readLine()) != null) {
+                    response.append(inputLine);
+                }
+                in.close();
```
You should put the reader in the try -> `try (BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));) {` Then this close can be removed. It's also saver because the close should be called in the finally block.

Hilbrand commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/jsonResponse/wso2iotsResponse.java:
```diff
@@ -0,0 +1,27 @@
+ * @author Ramesha Karunasena - Initial contribution
+ */
+
+public class wso2iotsResponse {
```
Classname should start with uppercase

Hilbrand commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/wso2iotsConfiguration.java:
```diff
@@ -0,0 +1,25 @@
+ * @author Ramesha Karunasena - Initial contribution
+ */
+
+public class wso2iotsConfiguration {
```
Classname should start with uppercase

RameshaKaru commented:
@Hilbrand  thanks for the feedback and I did the suggested changes.

RameshaKaru commented:
@martinvw  @ppieczul I have added the bridge to make the binding more feasible for future evolution, as you suggested.
But there isn't any specific API implemented in WSO2 IoT server to check whether it is online. It can be done by calling a device plugin specific API. But since plugins are represented as things, calling a device plugin specific API is not suitable at the bridge level. Therefore I have checked only the configuration validity for the bridge online state. However, it is possible to implement checking the server online status once that feature is available in the WSO2 IoT server in future.

Suggestions and feedback are highly appreciated.

martinvw commented:
> But there isn't any specific API implemented in WSO2 IoT server to check whether it is online. It can be done by calling a device plugin specific API. But since plugins are represented as things, calling a device plugin specific API is not suitable at the bridge level. Therefore I have checked only the configuration validity for the bridge online state. However, it is possible to implement checking the server online status once that feature is available in the WSO2 IoT server in future.

But I suppose you might need to login, or maybe request an overview of available devices or plugins, it that is succesfull then the bridge can be considered online. Otherwise you can choose to leave it unknow until one of the things fails or succeeds to communicate with the bridge for its own purpose. Please test this part through.

RameshaKaru commented:
@martinvw  I am writing a HealthCheck API for the WSO2 IoT Server to check whether it is online. Once it is completed I can call it from the binding to check online state.

RameshaKaru commented:
There was a way to check the online state of the WSO2 server without access tokens. I added it to the bridge handler.

But there is a small issue in my code: when the bridge status changes to online, the thing status also changes to online even when the thing is offline. (Then when thing handler is called after the given refresh period the thing changes back to offline. But in between, the thing is shown as online). Can you please tell me where I have done the mistake?

ppieczul commented:
@RameshaKaru thing handlers get notified on bridge status changes in method bridgeStatusChanged. Default implementation will update thing status to match bridge status. Try to override this method.

martinvw commented:
@RameshaKaru @ppieczul FYI there is a pull request in smarthome to change this behaviour and not put it online by default: https://github.com/eclipse/smarthome/pull/4086

RameshaKaru commented:
Did overriding bridgeStatusChanged

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/Wso2iotsBindingConstants.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
Can you update the copyrights?

`mvn license:format` 

Please commit only the ones for your binding.

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        if (validConfig) {
+            updateStatus(ThingStatus.OFFLINE);
```
Why update to OFFLINE here?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // No channels - nothing to do
```
You should handle the REFRESH command, can you?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+        if (refreshJob == null || refreshJob.isCancelled()) {
+            Runnable runnable = () -> {
+                try {
+
```
Please kill some newlines

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+                    checkConnection();
+
+                } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+        String errormsg = null;
+        String response1 = null;
+        BridgeConfiguration config = getConfigAs(BridgeConfiguration.class);
+        String hostname = config.hostname;
```
There is also a port configured, why do you ignore that?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+    private void checkConnection() throws Exception {
+
+        String errormsg = null;
+        String response1 = null;
```
What does this mean? And why do you declare them in this scope?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+        String url = "http://" + hostname + ":9763/services/Version.VersionHttpEndpoint";
+
+        URL obj;
```
Please join with next line and give a proper name.

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+        obj = this.validateCertificate(url);
+
+        if (obj == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
```
Is it possible to pass an hint about what went wrong?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+        } else {
+
+            HttpURLConnection con = (HttpURLConnection) obj.openConnection();
```
We have some proven technology for this in eclipse smarthome, please use that instead. 

https://www.eclipse.org/smarthome/documentation/javadoc/org/eclipse/smarthome/io/net/http/HttpUtil.html

Note that a lot of potential corner cases with closing connections and handling exceptions exist in such communication and are handled properly by the util.

Note that it also contains a builder like structure which is less cryptic, see an example in https://github.com/openhab/openhab2-addons/blob/116ba76b3de9b1d5e166d64b85f7272cc7bfb3f2/addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/ICloudConnection.java

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, errormsg);
+
```
Do no use these additional new lines.

It would be great if you could also review the report of [the Static code analysis](https://docs.openhab.org/developers/development/bindings.html#static-code-analysis), this also saves me some time while reviewing.

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+        try {
+            SSLContext sc = SSLContext.getInstance("SSL");
+            sc.init(null, trustAllCerts, new java.security.SecureRandom());
+            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
```
Please don't you might impact / compromise other bindings, if you need you might need to wait for https://github.com/openhab/openhab2-addons/pull/3919 and the related ESH PR to be completed.

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+
+    private ScheduledFuture<?> refreshJob;
+
+    private BigDecimal iotsResponseTemp;
```
Why would you store as `BigDecimal`?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+        logger.debug("Initializing thing handler.");
+
+        BuildingMonitorConfiguration config = getConfigAs(BuildingMonitorConfiguration.class);
+        logger.debug("config apikey = (omitted from logging)");
```
👍

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+            validConfig = false;
+        }
+        bridgeHandler = getBridgeHandler();
+        if (bridgeHandler == null || !bridgeHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
```
ThingStatus is an `Enum` can be compared with `==`

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        bridgeHandler = getBridgeHandler();
+        if (bridgeHandler == null || !bridgeHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
+            logger.debug("Bridge handler not found or not ONLINE.");
```
You log this twice, from line 127 / 134 and here

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        bridgeHandler = getBridgeHandler();
+        if (bridgeHandler == null || !bridgeHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
```
`Enum` see above.

martinvw replied:
@htreu I always forget do you need this ONLINE check of does ESH already do this?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+                        updateChannel(Wso2iotsBindingConstants.MOTION, iotsResponseMotion);
+                        updateChannel(Wso2iotsBindingConstants.HUMIDITY, iotsResponseHumidity);
+
+                    } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+
+        URL obj;
+
+        obj = this.validateCertificate(url);
```
How does this validate the certifcate?

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+    private BigDecimal updateBuildingMonitorData(String sensorType, BuildingMonitorConfiguration config,
+            BridgeConfiguration configBridge) throws Exception {
+
+        BigDecimal iotsResponse = null;
```
I would prefer if you would not declare like this its not really Java like :-)

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/handler/BuildingMonitorHandler.java:
```diff
@@ -0,0 +1,387 @@
+            return null;
+        } else {
+
+            HttpURLConnection con = (HttpURLConnection) obj.openConnection();
```
Please consider HttpUtil / HttpRequestBuilder

martinvw commented on addons/binding/org.openhab.binding.wso2iots/src/main/java/org/openhab/binding/wso2iots/internal/Wso2iotsHandlerFactory.java:
```diff
@@ -0,0 +1,61 @@
+
+import static org.openhab.binding.wso2iots.Wso2iotsBindingConstants.*;
+
+//import java.util.Collections;
```
please remove

martinvw commented on addons/binding/org.openhab.binding.wso2iots/.classpath:
```diff
@@ -0,0 +1,10 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/main/java"/>
+	<classpathentry kind="lib" path="/home/rameshakarunasena/Downloads/gson-2.6.2.jar"/>
```
This seems local configuration?

htreu replied:
Also, GSON is provided in the targetplatform.

martinvw commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,33 @@
+            <parameter name="hostname" type="text" required="true">
+                <label>Host Name</label>
+                <description>Host name of the WSO2 IoTS</description>
+                <default>localhost</default>
```
Please add:

```
<context>network-address</context>
```

martinvw commented on addons/binding/org.openhab.binding.wso2iots/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+		<label>Temperature</label>
+		<description>Temperature in Celsius degrees</description>
+		<category>Temperature</category>
+		<state readOnly="true" pattern="%.1f °C" />
```
Consider implementing support for Units of Measurment, see a lot of recent code.

martinvw commented:
Hello @RameshaKaru, what is the status of this PR, do you think you will be able to proces the review comments?

Hilbrand commented:
Closing due to inactivity.
